ll2_global_planner/src/ll2_global_planner_nodelet.cpp

```diff
@@ -44,10 +49,178 @@ void Ll2GlobalPlannerNl::loadParams()
 
 void Ll2GlobalPlannerNl::laneletMapCb(const autoware_lanelet2_msgs::MapBin& map_msg)
 {
+  lanelet_map_ = std::make_shared<lanelet::LaneletMap>();
+  lanelet::utils::conversion::fromBinMsg(map_msg, lanelet_map_);
+  initialized_ = true;
+  ROS_INFO("Loaded Lanelet map");
 }
 
 void Ll2GlobalPlannerNl::poseGoalCb(const geometry_msgs::PoseStamped::ConstPtr& pose_msg)
 {
+  if (!initialized_)
+  {
+    ROS_WARN("Waiting for lanelet2 map");
+    return;
+  }
+
+  planRoute(pose_msg->pose.position);
+}
+
+void Ll2GlobalPlannerNl::planRoute(const geometry_msgs::Point& goal_point)
+{
+  const float duplicate_dist_threshold = 0.1;
+
+  // Find the nearest lanelet to the goal point
+  BasicPoint2d local_point(goal_point.x, goal_point.y);
+  Lanelet goal_lanelet = getNearestLanelet(local_point);
+
+  // Get the current vehicle position
+  geometry_msgs::TransformStamped tf_msg;
+  try
+  {
+    tf_msg = tf_buffer_.lookupTransform("map", "base_link", ros::Time(0), ros::Duration(2.0));
+  }
+  catch (tf2::TransformException &ex)
+  {
+    ROS_WARN_THROTTLE(2, "%s", ex.what());
+    ROS_WARN_THROTTLE(2, "Waiting for map -> base_link transform");
+    return;
+  }
+
+  BasicPoint2d starting_point(
+    tf_msg.transform.translation.x, tf_msg.transform.translation.y);
+  const Lanelet& starting_lanelet = getNearestLanelet(starting_point);
+
+  // Plan a route from current vehicle position
+  traffic_rules::TrafficRulesPtr traffic_rules{traffic_rules::TrafficRulesFactory::instance().create(Locations::Germany, Participants::Vehicle)};
+  routing::RoutingGraphUPtr graph = routing::RoutingGraph::build(*lanelet_map_, *traffic_rules);
+  Optional<routing::LaneletPath> shortest_path_opt = graph->shortestPath(starting_lanelet, goal_lanelet);
+  routing::LaneletPath shortest_path;
+
+  if (!shortest_path_opt)
+  {
+    ROS_WARN("Could not find path in lanelet map!");
+    return;
+  }
+
+  shortest_path = shortest_path_opt.value();
+  LaneletSequence continuous_lane = shortest_path.getRemainingLane(shortest_path.begin());
+
+  if (continuous_lane.size() != shortest_path.size())
+  {
+    ROS_WARN("This route contains a lane change which is currently unsupported");
+    return;
+  }
+
+  ROS_INFO("Found a path containing %lu lanelets", shortest_path.size());
+
+  // Convert to autoware waypoints
+  std::vector<autoware_msgs::Waypoint> waypoints;
+
+  // Loop over each lanelet
+  for (auto& lanelet : continuous_lane.lanelets())
+  {
+    const std::string turn_direction = lanelet.attributeOr("turn_direction", "straight");
+    traffic_rules::SpeedLimitInformation speed_limit = traffic_rules->speedLimit(lanelet);
+    ConstLineString3d centerline = lanelet.centerline();
+    int wp_length = centerline.size() - 1;
+
+    // Loop over each centerline point
```

Comment:
Lanelet centerline in general can be very sparse (ie long straight road may only have 2 points in center). Does planner/controller assume that waypoints are more regularly spaced?

Reply:
Ahh, I see what is happening. The lanelet2 map loader for Autoware.AI is loading the map and calling a function from lanelet2_extension utilties to overwrite the centerline with a fine resolution (1m).

Thus when global planner calls the centerline function it returns fine space points. 

I guess ideally, the global planner should not assume this function has been called on a lanelet2 map. However for Autoware.AI I guess it will always be called, so perhaps just a comment that it expects centerline to be fine resolution.

Reply:
I agree that this should be documented - preferrably in the global planner.

Reply:
Thanks @simon-t4 for tracking that down! I either didn't know that myself or must have completely forgot :smile: 

I agree, I'll add a comment above the use of the `centerline()` function.